onmt/transforms/terminology.py

```diff
@@ -0,0 +1,346 @@
+        term_counter = 0
+
+        max_terms_allowed = int(len(tokenized_source) * self.term_example_ratio)
+        is_match = False
+        for match_end, pair in self.automaton.iter_long(lemmatized_source_string):
+
+            if term_counter == max_terms_allowed:
+                break
+
+            match_start = match_end - len(pair[0]) + 1
+
+            # We ensure that the target lemma is present in the lemmatized
+            # target string, that the match is an exact match (there is
+            # whitespace before or after the term)
+            # and we perform some bound checking.
+            if (
+                (pair[1].lower() not in " ".join(lemmatized_target).lower())
+                or (
+                    len(lemmatized_source_string) != match_end + 1
+                    and not (lemmatized_source_string[match_end + 1].isspace())
+                )
+                or (
+                    not lemmatized_source_string[match_start - 1].isspace()
+                    and match_start != 0
+                )
+            ):
+                continue
+            else:
+                term_counter += 1
+
+                # Map the lemmatized string match index to
+                # the lemmatized list index
+                lemma_list_index = 0
+                for i, w in enumerate(lemmatized_source):
+                    if lemma_list_index == match_start:
+                        lemma_list_index = i
+                        break
+                    else:
+                        lemma_list_index += len(w) + 1
+
+                # We need to know if the term is multiword
+                num_words_in_src_term = len(pair[0].split())
+                src_term = " ".join(
+                    tokenized_source[
+                        lemma_list_index : lemma_list_index + num_words_in_src_term
+                    ]
+                ).strip()
+
+                # Join multiword target lemmas with a unique separator so
+                # we can treat them as single word and not change the indices.
+                tgt_term = pair[1].replace(" ", "∥").rstrip().lower()
+                source_with_terms.append(
+                    f"{lemmatized_source_string[offset: match_start]}"
+                    f"{self.src_term_stoken}∥{src_term}∥{self.tgt_term_stoken}∥"
+                    f"{tgt_term}∥{self.tgt_term_etoken}"
+                )
+
+                offset = match_end + 1
+                is_match = True
+
+        if is_match:
+            source_with_terms.append(lemmatized_source_string[offset:])
+            tokenized_source_with_terms = "".join(source_with_terms).split()
+
+            if not (
+                len(tokenized_source)
+                == len(lemmatized_source)
+                == len(tokenized_source_with_terms)
+            ):
+                final_string = " ".join(tokenized_source)
+                fixed_punct = re.sub(r" ([^\w\s｟\-\–])", r"\1", final_string)
```

Comment:
These "fixed_punct" subs look weird, why is this necessary? Side effect of the term insertion?

Reply:
Yes, indeed, these are side effects due to spacy's tokenization/lemmatization.